### DISCUSSION TITLE
Bluetooth: ISO: Host package sequence number.

### DISCUSSION THREAD

**Discussion by Thalley** [zephyrproject-rtos/zephyr#43317] [RFC]
The current implementation (as of SHA 0e8f556b1765d077d80ed12fc3b8317ec76c2eb9) is that the sequence number is just naively incremented per SDU send from the application/host to the controller for the specific ISO stream. 

The core spec (Version 5.3, Vol 6, Part G, Section 2 ISOAL Features) mandates that the sequence number is incremented once per SDU interval, which is current not the case. 

So we need to increment the SDU interval via a timer in the host, started when the ISO channel is connected. This applies for both unicast and broadcast. This change is pretty trivial, and won't require much discussion (if any). 

Furthermore, to keep the timer in sync with the timer in the controller, as well as the sequence number, we need to use the `LE Read ISO TX Sync` HCI command. This will need to be called periodically either by the application, or by the host.
There is no way for the controller to indicate to the host that the timestamps/sequences numbers are out of sync. 

Below will be a few discussion points in comments, with different ways of implementing this sync between the host and the controller.

**Comment by Thalley** [comment]
**Sequence number increment: Solution 1**

In this solution, the host will just naively increment the sequence number once per SDU interval using a timer. This is a very simple solution and requires little to no thought. 

The side effect of this, is that if the application attempts to send 2 SDUs in a single SDU interval on a single stream, they will get the same package sequence number (unless we choose to reject the send request), and when sent to the controller won't be sent on air. Since the controller cannot reject it, it is best if the host rejects it if the application sends two SDUs in a single SDU interval on a single stream. 

The application will _not_ be able to enqueue multiple packets in the controller at the same time

**Comment by Thalley** [comment]
**Sequence number increment: Solution 2**

In this solution, the host will  increment the sequence number once per SDU interval using a timer, but also once per send request from the application. 

The application will be able to enqueue multiple packets in the controller (with different sequence numbers). 

The implementation will be slightly more complex than https://github.com/zephyrproject-rtos/zephyr/discussions/43317#discussioncomment-2272818 since we need to figure out when to increment the sequence from the timer and when we need to increment it from application requesting a send.

**Comment by Thalley** [comment]
**Sequence number increment: Solution 3**

In this solution we simply offload the sequence number to the application. Add another field in the `bt_iso_chan_send` to set the sequence number. This will be very simple for the host, but will require more from the application. The application will also be responsible for syncing the sequence number between the host/application and the controller. 

This solution is however also the most flexible, as it will allow an application to schedule ISO SDUs in the future (e.g. sending a SDU with a sequence number 10 above the current sequence number will schedule it 10 * SDU interval in the future. 

The application will however require that the TS of the last sent message is also known to the application.

**Reply by Casper-Bonde-Bose** [reply]
I think this is the right solution - also remember that the application shall generate one SDU per SDU interval - hence in the ideal world the two will always be in sync - and if they are not I do not think forcing a sequence number update is the right solution. (so actually this is what we have today)
The only case where we might want to have the host stack increment the timer would be in the case where the SDU stream stops for a longer period(silent period of audio) - in that case it would make sense for the host stack to handle the increments. Hence perhaps a mix of solution 2 and 3. 
To handle host transport jitter we will need to enqueue at least one extra SDU in the controller buffers - hence initially send #1 and #2 at the same time.
Maybe we need some input from the controller developers on this. I would think that the sequence number is only needed to do CIG PDU synchronization - ensuring the frames that goes into each CIS represent the same audio content (but could be different channels). There is a part in the core specification that allows the ISOAL upper layer to synchronize its sequence number to the sequence number of the incoming data - which suggest the sequence number of the upper to lower direction should not be used to convert to actual timing.

For devices that have access to the Bluetooth controller subsystem timing will have no problem using a timer for this, but in case of the host running on a separate CPU (for example using UART transport) this is not possible - over time the system clock on the host will drift relative to the BT controller - and even on a peripheral device with access the the Bluetooth controller timing you would need a hook into the lower layers of the controller to get peripheral-central clock drift information as the clock accuracy of the system allows 250ppm clock inaccuracy, which corresponds to one SDU (audio frame) every 40 seconds I believe.

I think the only way for a host/application to do adjustments for this drift is to use the timing of the data transmitted events (number of complete packets hci events). Average timing of this event will correlate to the central Bluetooth device timing, and this time reference can be used by the audio subsystem to do the mandatory sample rate adaption.

In the case where ISOAL framed PDU's are used I believe the idea is to let the host layer handle timing and offload handling of the drift to the other side of the link - in this case drift is allowed and a host timer could be used to generate SDU's and sequence number increments - but we still need to handle jitter - I would suggest at least one SDU interval of silence should past before automatically incrementing the sequence number without receiving a PDU.

The consequence of wrongly incrementing the sequence number in the host stack would be to introduce an audio gap - which really should only occur as a result of bad RF. (here I mean incrementing when there is still data in the controller buffer) .

The last aspect of this is what to do when a SDU is missing and a timer is used to increment the sequence number. If the host do nothing the sink side will not know what is going on, and will repeat attempting to receive the SDU. If in stead an empty PDU is sent to the controller the sink side will see a SDU and know that no more data will be sent. also the use of empty SDU's will allow the flow of number-of-complete-packet events to keep flowing - to keep time synchronization running.

**Reply by Thalley** [reply]
Thanks for your input. I have a couple comments: 

> remember that the application shall generate one SDU per SDU interval 

Do you have a reference for this? Because I have not seen this requirement, but rather in stead the core spec has this paragraph which I think says the opposite: 
> The sequence number shall be incremented for each SDU_Interval, whether or not an SDU was received from or sent to the upper layer.

> The only case where we might want to have the host stack increment the timer would be in the case where the SDU stream stops for a longer period(silent period of audio) - in that case it would make sense for the host stack to handle the increments. Hence perhaps a mix of solution 2 and 3.

Interesting thought. We could have the host maintain a sequence number based on a timer which will act as a minimum sequence number, while still allowing the application to set the sequence number in the send API, and reject in case the sequence number is lower than the host maintained value. Is that similar to what you had in mind? 

> For devices that have access to the Bluetooth controller subsystem timing will have no problem using a timer for this, but in case of the host running on a separate CPU (for example using UART transport) this is not possible - over time the system clock on the host will drift relative to the BT controller - and even on a peripheral device with access the the Bluetooth controller timing you would need a hook into the lower layers of the controller to get peripheral-central clock drift information as the clock accuracy of the system allows 250ppm clock inaccuracy, which corresponds to one SDU (audio frame) every 40 seconds I believe.
> 
> I think the only way for a host/application to do adjustments for this drift is to use the timing of the data transmitted events (number of complete packets hci events). Average timing of this event will correlate to the central Bluetooth device timing, and this time reference can be used by the audio subsystem to do the mandatory sample rate adaption.

Regarding drift, if we let the application call and get the result of the `LE Read ISO TX Sync` HCI command, could the application not handle drift in that case? Assuming of course that we still enqueue at least 1 packet in the controller?

> The consequence of wrongly incrementing the sequence number in the host stack would be to introduce an audio gap 

Even worse; if the sequence number from the host is higher than the one maintained in the controller, the controller may discard all SDUs, causing all audio data to stop with no way of informing the host. 

> If in stead an empty PDU is sent to the controller the sink side will see a SDU and know that no more data will be sent. also the use of empty SDU's will allow the flow of number-of-complete-packet events to keep flowing - to keep time synchronization running.

Empty PDU will (may?) not generate an SDU and the remote host will not be notified, and only SDUs (e.g. with 0 bytes of payload) will generate an SDU to the remote host IIRC.

**Reply by Casper-Bonde-Bose** [reply]
> Do you have a reference for this? Because I have not seen this requirement, but rather in stead the core spec has this paragraph which I think says the opposite:

I do not think the specification needs to state this - if the application do not generate exactly one SDU every SDU interval then you have gaps in your audio stream or buffer overflow over time - hence a broken system :-)

> Interesting thought. We could have the host maintain a sequence number based on a timer which will act as a minimum sequence number, while still allowing the application to set the sequence number in the send API, and reject in case the sequence number is lower than the host maintained value. Is that similar to what you had in mind?

I think I would do it as a guard timer starting when the stream is enabled. It could be one or two SDU intervals (kconfig value). The timer is reset for each SDU written by the application hence if it fires it is because a SDU was not written in time and the host stack could increment based on this. Actually we do not need a timer if we just store a timestamp for each SDU send - then if there is a gap above a certain threshold when a new SDU is send we can increment the sequence number accordingly - the timer is only needed if we want to transmit empty PDU's to indicate missing data.
As we do not have accurate timing I think it is impossible to calculate an exact SDU sequence number based on timing over a long period - 250ppm would cause your calculations to be wrong by one every 40 seconds - hence in a 2 hour movie you would be very wrong even if your clock accuracy was lower.

> Regarding drift, if we let the application call and get the result of the LE Read ISO TX Sync HCI command, could the application not handle drift in that case? Assuming of course that we still enqueue at least 1 packet in the controller?

It will not be very accurate but yes over a long period it could be used.

> Even worse; if the sequence number from the host is higher than the one maintained in the controller, the controller may discard all SDUs, causing all audio data to stop with no way of informing the host.

No it should not. If the host do not apply a timestamp in the SDU's to be send the controller is responsible to schedule transmission based on when in time the SDU was received - which is why I doubt any controller implementation will use the sequence number for anything but aligning data between multiple CIS'es. and maybe in case of a gap of one skip a CIS event:
> When an HCI ISO Data packet sent by the Host does not contain a Time_-
 Stamp or the Time_Stamp value is not based on the Controller's clock, the
 Controller should determine the isochronous event to be used to transmit the
 SDU contained in that packet based on the time of arrival of that packet

> Empty PDU will (may?) not generate an SDU and the remote host will not be notified, and only SDUs (e.g. with 0 bytes of payload) will generate an SDU to the remote host IIRC.

Over air there are two types of "empty" packets. A NULL packet (NPI bit set) indicates there is no data - hence the receiving controller will not see this as a PDU and will require use of retransmissions for this missing PDU and you will start to congest retransmissions until you reach the flush point - and then when data is actually to be sent you will be in retransmission congested state with only a few retransmissions left (I have a few slides of this in case you want a call for me to explain). If however empty packets are sent in replacement of missing data you will keep the system in normal state - having all buffers at the correct level for retransmission to be used for real data. It is not important if these empty PDU's are sent to the host on the sink side (but they should to kick in the PLC) the important part is the the buffer level in the controller are kept "intact".

**Reply by Thalley** [reply]
> > Do you have a reference for this? Because I have not seen this requirement, but rather in stead the core spec has this paragraph which I think says the opposite:

 > I do not think the specification needs to state this - if the application do not generate exactly one SDU every SDU interval then you have gaps in your audio stream or buffer overflow over time - hence a broken system :-)

The host/application can generate multiple SDUs in a single interval with increasing packet sequence number to enqueue them in the controller, e.g. in a 10ms (example SDU interval) it could generate SDU_1 (SN 1), SDU_2 (SN 2), SDU_3 (SN 3), etc. and enqueue all of them in the controller in a burst and the controller will then send them 1 by 1. In the case that the controller supports e.g. 10 packets, the host/app can enqueue 10 packets within a 10ms time frame, have a pause for 90-100ms and enqueue 10 more. 
Furthermore what is broken about an ISO stream that e.g. doesn't have any SDU for e.g. a second? (Please keep in mind that ISO != Audio, and that there are other usecases for it besides audio). 

> > Interesting thought. We could have the host maintain a sequence number based on a timer which will act as a minimum sequence number, while still allowing the application to set the sequence number in the send API, and reject in case the sequence number is lower than the host maintained value. Is that similar to what you had in mind?

> I think I would do it as a guard timer starting when the stream is enabled. It could be one or two SDU intervals (kconfig value). The timer is reset for each SDU written by the application hence if it fires it is because a SDU was not written in time and the host stack could increment based on this. Actually we do not need a timer if we just store a timestamp for each SDU send - then if there is a gap above a certain threshold when a new SDU is send we can increment the sequence number accordingly - the timer is only needed if we want to transmit empty PDU's to indicate missing data.

If the host/app needs to increment the sequence number once per SDU interval, I don't really see any way to avoid a timer, unless we want to call the `LE Read ISO TX Sync` HCI command before sending to do a quick sync-up before sending. 

 
> > Even worse; if the sequence number from the host is higher than the one maintained in the controller, the controller may discard all SDUs, causing all audio data to stop with no way of informing the host.

 > No it should not. If the host do not apply a timestamp in the SDU's to be send the controller is responsible to schedule transmission based on when in time the SDU was received - which is why I doubt any controller implementation will use the sequence number for anything but aligning data between multiple CIS'es. and maybe in case of a gap of one skip a CIS event:
> When an HCI ISO Data packet sent by the Host does not contain a Time_-
> Stamp or the Time_Stamp value is not based on the Controller's clock, the
> Controller should determine the isochronous event to be used to transmit the
> SDU contained in that packet based on the time of arrival of that packet
 
I think @cvinayak is more suited to answer this one. 

> > Empty PDU will (may?) not generate an SDU and the remote host will not be notified, and only SDUs (e.g. with 0 bytes of payload) will generate an SDU to the remote host IIRC.

> Over air there are two types of "empty" packets. A NULL packet (NPI bit set) indicates there is no data - hence the receiving controller will not see this as a PDU and will require use of retransmissions for this missing PDU and you will start to congest retransmissions until you reach the flush point - and then when data is actually to be sent you will be in retransmission congested state with only a few retransmissions left (I have a few slides of this in case you want a call for me to explain). If however empty packets are sent in replacement of missing data you will keep the system in normal state - having all buffers at the correct level for retransmission to be used for real data. It is not important if these empty PDU's are sent to the host on the sink side (but they should to kick in the PLC) the important part is the the buffer level in the controller are kept "intact".

My understanding is that only SDUs are sent to the host. An empty PDU which is not caused by the remote host (it can't really), should not reach the local host. 


Generally I think there's still a lot of confusion regarding the sequence numbers and how they should be used (the spec is also not being clear on this).

**Reply by Casper-Bonde-Bose** [reply]
> The host/application can generate multiple SDUs in a single interval with increasing packet sequence number to enqueue them in the controller...

True, that is why I added "over time" :-) so in average you would need one SDU every SDU interval - for audio. The system also needs to handle BN>1.

> Please keep in mind that ISO != Audio, and that there are other use-cases for it besides audio.

True good point - but the buffering and synchronization issue still applies unless FT=1 and you do not care about power consumption.
CIS'es are designed to periodically transfer synchronized data and uses the CIS timing for synchronization and numbering of the received PDU's - hence you cannot use a CIS as  you would use a normal ACL and attempt to transfer data at random points in time. (you can but it is not efficient). The correct way to use a CIS is to provide data at each SDU interval - even if most are empty - this will allow the central device to close the CIS event early and not force it to use all CIS subevents. In case the data provider is the central device, then I think the controller could close the CIS event early, but not sure all will - at least I have seen one controller that keeps the CIS event running "hoping" to get the missing data from the host before the CIS event runs out.

> My understanding is that only SDUs are sent to the host. An empty PDU which is not caused by the remote host (it can't really), should not reach the local host.

I disagree. The HCI packet description have information in the header to allow a host to drive a codec with PLC (for example data received with error) - if the controller do not issue ISO data events for every SDU interval then it overcomplicates the audio processing and a host is forced to implement quite a large buffer to do timer based handling of this (again the different time domains forces handling of jitter) - hence adding to the overall latency.
The host should be notified when:
- data is received correctly (obviously)
- when data is flushed for example due to reception errors (this include reception of NULL packets where SDUs should have been present)
- when empty packets are received (no payload but NPI bit not set)
Unfortunately I have not found a shall in the specification on this :-) 

Why do you say "it can't really" - I do not see any statement that ISO_SDU_Length cannot be 0 (section 5.4.5 HCI ISO Data packets).
Of cause a controller could implement this mechanism - sending an empty PDU (PB-flag set to complete SDU) if the TX buffer is empty, which would solve the retransmission issue.

> Generally I think there's still a lot of confusion regarding the sequence numbers and how they should be used (the spec is also not being clear on this).

How about doing the empty packet approach and use the number of complete packets events to drive the correct sequence number update. So as @asbjornsabo suggested let the data write from an application drive the sequence number increment when there is a steady flow of data. The host knows how many buffers are in use in the controller - and when this runs below a certain threshold the host stack will start injecting empty packets every time it receives a number of complete packet event and the buffer level in the controller is still below the threshold - when the application have provided new data the buffer level will not be below the threshold.
This will solve both the sequence number increment requirement, the buffer level issue as well as the time-domain challenge. But it will force the host to use at least 1 TX buffers in the controller - hence it shall provide a new SDU before it received the number-of-complete-pkt event from the previous.

**Reply by Thalley** [reply]
> hence you cannot use a CIS as you would use a normal ACL and attempt to transfer data at random points in time. (you can but it is not efficient). The correct way to use a CIS is to provide data at each SDU interval 

ISO (specifically CIS) have some features that are interesting to explore for non-periodic data, e.g. to handle very low intervals (less than the minimum 7.5ms ACL interval), as well as establishing multiple CIS towards a single device to send data on e.g. 4 streams in a single interval. While I agree that this perhaps is misusing the intention of ISO, there are certainly use cases where misusing ISO/CIS can provide functionality not previously possible with BT. Since the core spec does not prohibit such behavior, I'd rather avoid the ISO API to only support strictly periodic data. 

> Why do you say "it can't really" - I do not see any statement that ISO_SDU_Length cannot be 0 (section 5.4.5 HCI ISO Data packets). 

That is true. It isn't written clearly in any sentence, but consider this part: 
![image](https://user-images.githubusercontent.com/2520413/156556115-1c0b98e7-5d61-4f3f-95ec-bd819754b27e.png)
The PB_Flag shall be set, and has 1 of 4 values, each of which describes that the data contains and SDU or part of an SDU. If no SDU is sent from the remote host to the remote controller, there is no SDU being sent, and thus this flag will be in conflict in my opinion. I don't think the controller is allowed to send SDUs on behalf of the host. In other words, a remote host may send an SDU with length = 0, but that will be an SDU nonetheless, but if the remote host does not send an SDU (which is isn't required to do for each interval), then no SDU should be sent to local host, as that would indicate that either of the controllers are generating a SDU on behalf of the host. Thus I think that 

> when empty packets are received (no payload but NPI bit not set)

that are sent by the remote host, should not be sent to the local host. 

> How about doing the empty packet approach and use the number of complete packets events to drive the correct sequence number update. So as @asbjornsabo suggested let the data write from an application drive the sequence number increment when there is a steady flow of data. The host knows how many buffers are in use in the controller - and when this runs below a certain threshold the host stack will start injecting empty packets every time it receives a number of complete packet event and the buffer level in the controller is still below the threshold - when the application have provided new data the buffer level will not be below the threshold.
This will solve both the sequence number increment requirement, the buffer level issue as well as the time-domain challenge. But it will force the host to use at least 1 TX buffers in the controller - hence it shall provide a new SDU before it received the number-of-complete-pkt event from the previous.

Interesting approach, and could indeed "fix" the incremental of the sequence number without using a timer. It will, however, send unnecessary many empty SDUs to the remote host (assuming that it doesn't get empty SDUs if the host doesn't send), which may or may not be good. In the case of a ISO stream with "breaks" in the e.g. audio (e.g. a walkie talkie implementation), the constant receive of SDUs will need to wake up the host processor (in multi-processor builds) with empty SDUs, just for the host to possibly ignore or discard them. I'm not strictly against this proposal (in fact I'll add it as a solution proposal now), but it will have side effects not found in the others, although waking the host processor to increment a value by a timer is roughly the same.

**Reply by Casper-Bonde-Bose** [reply]
The walkie talkie example is a good example as you have frequent silence periods. In case empty PDU's are not sent while silence the controller will need to service every single CIS sub-event - this means toggling on/off the BT receiver and transmitter throughout the entire CIS event exchanging nothing but NULL packets as well as forcing the radio CPU to stay awake for the entire CIS event. IMO this requires significantly more power than transferring a single empty PDU when there is no audio data to indicate it is OK for the central to close the CIS event early and allow the radio controller to go to sleep until the next CIS event. But it surely depends on the CIG configuration (NSE) and the cost of waking the host CPU - but unless we are talking a mobile phone or a PC with a big OS I doubt it is more expensive. Whether or not a controller allows a vendor specific command to disable empty ISO data events is implementation specific - but it should be a core-spec feature if you ask me.
In the walkie talkie example the application knows there will be no more data, hence can easily indicate that without the need to use guard timeout periods.

Note that over air a NULL packet and an empty SDU have the exact same length - the only difference is if the NPI bit is set or not. All the header content the host sees (sequence number timestamp etc.) is stripped away before transmission and recreated on the receiver side based on CIS timing. On an ACL you have a more data bit to indicate when to stop a connection event but on a CIS you use the knowledge about the number of expected PDU's to be received in each CIS event to determine when you can close the event.

I agree the controller should not generate the empty SDU's, but I'm just stating that it could to save power if the host do not provide a steady flow of SDU's.

**Comment by Thalley** [comment]
**TX Sync Solution 1**

To sync the timestamp and sequence number, the` LE Read ISO TX Sync` HCI command can be used. 

In this solution, the application is responsible for calling a function that sends this command to the controller, and then the resulting event can then be used for the host or application (depending on the Sequence number solution) to correct any out-of-sync sequence numbers. 

It is thus up to to the application to derive when it makes sense to call this, and how often.

**Reply by Casper-Bonde-Bose** [reply]
I do not see much value in this command on systems that do not have access to actually read the controller timestamp. I think the data transmitted events provide better timing accuracy (the flush timeout + a few ms) as you will get these events as soon as a SDU is transmitted where as the host would need to periodically send this HCI command hence you will have timing inaccuracies of twice the flush timeout - or the flush timeout + the periodic sample interval which ever is lowest.

The only case I can see where you would need to use this command is after a period of silence where you want to start off with the correct sequence number on your SDUs and have not implemented what you suggest above.

If you have access to the local BT controller timing you could use the command to determine clock drift between central and peripheral which have a lot of value to do adaptive sample rate conversion - but if you already have access to the Bluetooth timing I would assume there are better ways to read the clock drift that this command..

I think TX sync across multiple CISes are more important - ensuring the same audio frame is send across all CISes in a group - but I assume the controller will handle this if the host ensures they are tagged with the correct sequence number.

**Reply by Thalley** [reply]
> The only case I can see where you would need to use this command is after a period of silence where you want to start off with the correct sequence number on your SDUs and have not implemented what you suggest above.

In the case the host/app timer (or whatever triggers a `bt_iso_chan_send`) runs slower than the timer (or whatever) controlling the controller sequence number, then if the controller sequence number gets 1 value ahead of the host, it may discard all future packets. The only way that I see to avoid this is to call this `LE Read ISO TX Sync` HCI command to sync the sequence number, and thus probably increment the sequence number on the host/app. Given that the controller cannot indicate out-of-sync sequence numbers to the host, this will need to be called periodically. 

> I think TX sync across multiple CISes are more important - ensuring the same audio frame is send across all CISes in a group - but I assume the controller will handle this if the host ensures they are tagged with the correct sequence number.

This is important as well. I guess the worse case scenario is where you have 2 (or more) CIS, and the host/app (possibly due to drift) first sends an SDU on CIS_1 to the controller 1ms (or less) before the next SDU interval, and the next (given that we cannot do it in parallel) sends the SDU on CIS_2 to the controller 1ms (or less) __after__ the SDU interval, causing a SDU send on both streams to be sent in two different SDU intervals. However, I can only imagine this being an extremely rare occurrence, and if there's drift it will unlikely more than once in a row.

**Reply by Casper-Bonde-Bose** [reply]
> However, I can only imagine this being an extremely rare occurrence, and if there's drift it will unlikely more than once in a row.

Actually I have tested this on a system that ignores the sequence number - and it happens quite often if the first PDU is sent to the controller close to the ISO event as you mention - then I experience left/right sync shift back and forth one audio frame every 10-20 second. If however the first PDU is send with a reasonable distance to the anchor point it could play for minutes before it happened. 
Using the controller buffer to store at least one SDU removes the shift back and forth, but you still have the left/right offset by one audio frame - which breaks the specification - I believe the BT subsystem shall guarantee 50us left/right sync accuracy to be qualifiable. So the system shall have a means to synchronize audio frames between multiple CIS'es in a group.

**Comment by Thalley** [comment]
**TX Sync Solution 2**

To sync the timestamp and sequence number, the` LE Read ISO TX Sync` HCI command can be used. 

In this solution, the host will send this command to the controller periodically. This can either by configured by a KConfig in e.g. MS (e.g. send every 10000ms/10s), or relative to the SDU interval, e.g. send it every 100 SDU intervals. 

This solution can work together with https://github.com/zephyrproject-rtos/zephyr/discussions/43317#discussioncomment-2272874, i.e. having the host call this periodically, as well as allowing the application to call it manually.

**Comment by asbjornsabo** [comment]
For a correct supply of audio data from host to controller, the host must anyway send (at least on average) exactly one SDU per SDU interval. So "naively incrementing" should be fine.  That will fulfill the spec requirement of incrementing once per SDU interval.

So, I think the question is not so much about the numbering as it is about knowing when to send SDUs, i.e. keeping track of how many SDU intervals have elapsed and when send SDUs to the controller.  For a "live" system, that should be given by the flow of data from the A/D converter (which must exist somewhere).  For stored data (less common?) a timer is required, I believe.

**Reply by Thalley** [reply]
The naively incrementing won't work in case that the application doesn't send every SDU interval (which it isn't required to do). 

The numbering is indeed not important, but rather sending the correct number to the controller. 

Please keep in mind that we are discussing ISO data here, and not audio, and that any ISO data source may not provide data every SDU interval. 

For audio data, there may also be skips in time. E.g. a chat application (teams, discord, etc.) where there's a "Push to talk" (similar to walkie talkie), where you may not want to disconnect the iso/audio stream between each "session". In those cases, I don't think it would make to expect the application to send empty SDUs (empty in the sense that there is no audio data).

**Comment by Thalley** [comment]
**Sequence number increment: Solution 4**

In this solution the host will continuously send ISO packets to the controller, and use the command complete to increment the sequence number, rather than using a timer, when a specific ISO channel is considered "idle". This should ensure correctness for the sequence number, assuming that we can enqueue enough packets to never miss an interval. 

The side effects of this is that the controller will continuously send SDUs instead of empty PDUs, causing the remote host to receive empty SDUs continuously (which arguably is the intention of the ISO feature). A negative side effect is that the host will need to reserve at least 1 but more like 1-2 ISO TX buffers to implement this feature.

**Comment by Thalley** [comment]
It seems that the conclusion (based on numerous discussions) is that **TX Sync Solution 1** together with **Sequence number increment: Solution 3** is the solution. 

In this case, we basically offload everything to the application, and add the packet sequence number to the `bt_iso_chan_send` function, as welll as an optional timestamp parameter. This gives the maximum flexibility, which is useful for the different usecases that ISO supports. 

So basically we are going to change 
```c
int bt_iso_chan_send(struct bt_iso_chan *chan, struct net_buf *buf);
```
to 
```c
int bt_iso_chan_send(struct bt_iso_chan *chan, struct net_buf *buf, uint16_t sn, uint32_t ts);
```

For Audio data (sent via `bt_audio_stream_send`), we may want to omit `sn` and possibly also the `ts` and have BAP automatically send something every SDU interval (empty SDUs if nothing is sent by the higher layers), but that is TBD. 


@Casper-Bonde-Bose  Does this seem good to you?

**Reply by Thalley** [reply]
https://github.com/zephyrproject-rtos/zephyr/issues/43859 will be worked on soon.

**Comment by Thalley** [comment]
Closed as issue has been fixed
